StandardFiniteFieldsOSCAR.jl

```diff
@@ -1,443 +1,443 @@
+     m = div(4*q, 5)
+     while gcd(m,q) != 1
+         m -= 1
+     end
+     a = div(2*q, 3)
+     return mod((m*i + a), q)
+ end
+
+ # Given a field F and Steinitz number n, give the corresponding field element.
+ # we REQUIRE that F is a standard finite field TODO: using @assert?
+ function elementfromsteinitznumber(F::FinField, n)
+     p = characteristic(F)
+     q = order(F)
+     if n < 0 || n > q
+         error("We need to have 0 <= n <= q")
+     end
+     if n == 0
+         return zero(F)
+     elseif q == p
+         return F(n)
+     else
+         vectorrep = GF(p).(digits(n, base = Int(p)))
+         # this forms a linear combo of F.towervasis rows using vectorrep as coefficients,
+         # and then convert this vector to an element of F.
+         return F(vectorrep * @view F.towerbasis[1:length(vectorrep), :])
+     end
+ end
+
+ # Returns an element a in F that is NOT an rth root of unity
+ # we REQUIRE that F is a standard finite field TODO: using @assert?
+ function non_rth_root(F::FinField, r)
+     q = order(F)
+     if mod(q-1, r) == 0
+         i = 0
+         a = zero(F)
+         while a == 0 || a^(div(q-1,r)) == one(F)
+             i += 1
+             a = elementfromsteinitznumber(F, standardaffineshift(q, i))
+         end
+         return a
+     else
+         return nothing
+     end
+ end
+
+
+ function standardirreduciblecoefflist(F::FinField, r, a)
+     q = order(F)
+     l = zeros(F, Int(r)+1) # NOTE - AbstractAlgebra does not allow this be ZZRingElem
+     l[Int(r)+1] = one(F)
+     l[1] = a
+     l[2] = one(F)
+     # inc is the expected number of nonzero coefficients
+     inc = 1
+     while q^inc < 2*r
+         inc += 1
+     end
+     # allowing non-zero coeffs up to position d
+     # after every r attempts allow inc more non-zero coeffs
+     d = 0
+     count = 0
+     qq = 0
+     while !is_irreducible(polynomial(F, l))
+         if mod(count, r) == 0 && d < r-1
+             d += inc
+             if d >= r
+                 d = r-1
+             end
+             qq = q^(d-1)
+         end
+         # q can be very very large so Int is not big enough...
+         st = digits(standardaffineshift(qq,count), base = BigInt(q), pad = d-1)
+         # TODO: we can remove this while loop when digits bug for n = 0 is fixed
+         while length(st) < d-1
+             push!(st, 0)
+         end
+         for k in 2:d
+             l[k] = elementfromsteinitznumber(F, st[k-1])
+         end
+         count += 1
+     end
+     return l
+ end
+
+ # returns the Steinitz number corresponding to the polynomial g(X),
+ # where f = X^r + g(X) is the standard irreducible polynomial over FF(p, r^(k-1))
+ function steinitznumberforprimedegree(p,r,k)
+     Fp = standardfinitefield(p,1)
+     if !haskey(Fp.__attrs, :steinitzprimedegree)
+         Fp.steinitzprimedegree = Dict{ZZRingElem, Dict{ZZRingElem, ZZRingElem}}()
+     end
+     stpd = Fp.steinitzprimedegree
```

Comment:
Using regular attributes, this might look like this (untested)
```suggestion
     stpd = get_attribute!(Dict{ZZRingElem, Dict{ZZRingElem, ZZRingElem}}, Fp, :steinitzprimedegree)
```
or maybe more readable:
```suggestion
     stpd = get_attribute!(Fp, :steinitzprimedegree) do
         Dict{ZZRingElem, Dict{ZZRingElem, ZZRingElem}}()
     end
```